cli/npm/managed/resolvers/local/bin_entries.rs

```diff
@@ -3,7 +3,201 @@
+          if !self.seen_names.insert(bin_name.to_string()) {
+            self.has_collision = true;
+            self.seen_names.clear();
+          }
+        }
+        deno_npm::registry::NpmPackageVersionBinEntry::Map(entries) => {
+          for name in entries.keys() {
+            if !self.seen_names.insert(name.clone()) {
+              self.has_collision = true;
+              self.seen_names.clear();
+              break;
+            }
+          }
+        }
+      }
+    }
+
+    self.entries.push((package, package_path));
+  }
+
+  /// Finish setting up the bin entries, writing the necessary files
+  /// to disk.
+  pub(super) fn finish(
+    mut self,
+    snapshot: &NpmResolutionSnapshot,
+    bin_node_modules_dir_path: &Path,
+  ) -> Result<(), AnyError> {
+    if !self.entries.is_empty() && !bin_node_modules_dir_path.exists() {
+      std::fs::create_dir_all(bin_node_modules_dir_path).with_context(
+        || format!("Creating '{}'", bin_node_modules_dir_path.display()),
+      )?;
+    }
+
+    if self.has_collision {
+      // walking the dependency tree to find out the depth of each package
+      // is sort of expensive, so we only do it if there's a collision
+      sort_by_depth(snapshot, &mut self.entries);
+    }
+
+    let mut seen = HashSet::new();
+
+    for (package, package_path) in &self.entries {
+      if let Some(bin_entries) = &package.bin {
+        match bin_entries {
+          deno_npm::registry::NpmPackageVersionBinEntry::String(script) => {
+            let name = default_bin_name(package);
+            if !seen.insert(name) {
+              // we already set up a bin entry with this name
+              continue;
+            }
+            set_up_bin_entry(
+              package,
+              name,
+              script,
+              package_path,
+              bin_node_modules_dir_path,
+            )?;
+          }
+          deno_npm::registry::NpmPackageVersionBinEntry::Map(entries) => {
+            for (name, script) in entries {
+              if !seen.insert(name) {
+                // we already set up a bin entry with this name
+                continue;
+              }
+              set_up_bin_entry(
+                package,
+                name,
+                script,
+                package_path,
+                bin_node_modules_dir_path,
+              )?;
+            }
+          }
+        }
+      }
+    }
+
+    Ok(())
+  }
+}
+
+// walk the dependency tree to find out the depth of each package
+// that has a bin entry, then sort them by depth
+fn sort_by_depth(
+  snapshot: &NpmResolutionSnapshot,
+  bin_entries: &mut [(NpmResolutionPackage, PathBuf)],
+) {
+  enum Entry<'a> {
+    Pkg(&'a NpmPackageId),
+    IncreaseDepth,
+  }
+  let mut want = bin_entries
+    .iter()
+    .map(|(p, _)| p.id.clone())
+    .collect::<HashSet<_>>();
+  let mut seen = HashSet::new();
+  let mut depths: HashMap<NpmPackageId, u64> =
+    HashMap::with_capacity(want.len());
```

Comment:
We can probably avoid cloning here and use a reference for `NpmPackageId` instead?